module/errorhandler/erroraware.go

```diff
@@ -0,0 +1,167 @@
+//
+// ThrowIrrecoverable can be a drop-in replacement anywhere we have a context.Context likely
+// to support Irrecoverables. Note: this is not a method
+func ThrowIrrecoverable(ctx context.Context, err error) {
+	signalerAbleContext, ok := ctx.(IrrecoverableSignalerContext)
+	if ok {
+		signalerAbleContext.ThrowIrrecoverable(err)
+	}
+	// Be spectacular on how this does not -but should- handle irrecoverables:
+	log.Fatalf("Irrecoverable error signaler not found for context, please implement! Unhandled irrecoverable error %v", err)
+}
+
+////////////////////////////////////////////////
+// Integrating it w/ ReadyDoneAware & friends //
+////////////////////////////////////////////////
+
+// If we want to do it using interface composition (see module.Component), we will need to build on the existing interfaces
+// note the irrecoverable management needs to be:
+// - set up before the call to start (if the start itself meets an irrecoverable condition)
+// - not throw / return error itself, except to an enclosing context
+
+type Startable interface {
+	Start(runCtx IrrecoverableSignalerContext) error
+}
+
+// later, in RunComponent, the plumbing happens ...
+type Component interface {
+	Startable
+	module.ReadyDoneAware
+}
+type ComponentFactory func() (Component, error)
+
+func RunComponent(parentCtx IrrecoverableSignalerContext, componentFactory ComponentFactory, handler OnIrrecoverable) error {
+
+	// reference to per-run signals for the component
+	var component Component
+	var cancel context.CancelFunc
+	var done <-chan struct{}
+	var irrecoverables chan error
+
+	// Tells us:
+	// - how to get started
+	// - how to create a closure to program the handler with a continuation
+	restart := func() error {
+		var err error // startup error, should be handled out of band
+
+		component, err = componentFactory()
+		if err != nil {
+			return err // failure to generate the component, should be handles out-of-band because a restart won't help
+		}
+
+		// context used to run the component
+		var runCtx context.Context
+		runCtx, cancel = context.WithCancel(parentCtx)
+
+		// signaler used for irrecoverables
+		var signalingCtx IrrecoverableSignalerContext
+		irrecoverables = make(chan error)
+		signalingCtx = WithIrrecoverableSignal(runCtx, &IrrecoverableSignaler{irrecoverables})
+
+		if err = component.Start(signalingCtx); err != nil {
+			// failed to start component: this should not trigger a restart
+			return err
+		}
+		// Anywhere inside the component, we can use signalingCtx.ThrowIrrecoverable(err), if the types support it
+		// and ThrowIrrecoverable(signalingCtx, err) if not
+
+		// wait for Ready
+		<-component.Ready()
+
+		done = component.Done()
+		return nil
+	}
+
+	err := restart()
+	if err != nil {
+		return err // failure to start
+	}
+
+	for {
+		select {
+		case err := <-irrecoverables:
+			// shutdown the component,
+			cancel()
+			// wait until it's doneC
+			<-done
```

Comment:
While we are waiting for `done`, its possible the `parentCtx` could be cancelled.

In this case, we should abort waiting for shutdown to complete and exit the function. Otherwise, we will still proceed to the `handler(err, restart)` below even though the parent ctx is cancelled.

I think we just need an additional `select` statement wrapping this `<-done`.